File: estagio/conting/conting/mestre.c
/*-------- mestre.c ----------------------*/
/* -----------------Inicio da funcao MestreA--------------------*/
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "global.h"
#include "prototipos.h"
#include "pvm3.h"

void  MestreA(double ce[max],double cle[max],int impr){

	int  nsize1[63],nsize2[63],nsize3[63],nsize4[63]; //local
	//int  jklasif[nlmax];          // local
	int  k,i,j;
/*----------------  Declaracao relativa ao PVM   -------------------*/
	int knt,info, bufid;
	int icont0,icontr,kount,ip,nrcaso,nofinal;
	int nbase,nilhasoma,ndivsoma,nnconvsoma,nyconvsoma;
	int nitens3,nitens4,nitens5,nitens6;
	int msgtype2,msgtype3,msgtype4,msgtype5,msgtype6;
	int nndim, casoAnalizado;
/* ---------------------------------------
Rotina MestreA(): Modo Paralelo Assincrono
-----------------------------------------
-----------------------------------------------
 PVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVM
---------------------------------------------
--------------------------------
 Rotulos das mensagens [ types ]
------------------------------*/
	msgtype2 = 31;
	msgtype3 = 41;
	msgtype4 = 51;
	msgtype5 = 61;
	msgtype6 = 71;
	nbase = NPROC;
	nofinal = NPROC-1;
	nilhasoma = 0;
	ndivsoma = 0;
	nnconvsoma = 0;
	nyconvsoma = 0;
	/*------- Inicio do Processamento da Contingencia AC Lista Reduzida--------
		PVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVM
	PROCESSAMENTO PARALELO ASSINCRONO
	-----------------------------------------------
	------------------------------------------------
	O LOOP AC e Executado em Paralelo Assincrono: O
	Processador Assim que Termina seu Caso Pega  um
	Novo Caso para Executar
	------------------------------------------------
	O Controle e Feito Somente pelo No "0" que Tam-
	bem Executa Calculos AC Enquanto Espera
	----------------------------------------------*/
	if ( nbase == 1 ){
		for(k=0; k < ncrit; k++){
			casoAnalizado = klasif[k]; // caso a ser analisado
			//printf("Caso a ser analizado => %d",casoAnalizado);
			//printf("\n Ioutlocal em mestrea =  %d", ioutlocal);
			contingenciaAC(ce,cle,impr,casoAnalizado,k);// esta rotina faz chamadas da contingencia AC
		}
	}
	if ( nbase > 1 ){
	/*---- Preparando a Carga dos Processadores  ------*/
		if(nrc < nbase){
			nndim = nrc;
			printf("Toda a lista e menor do que o no. de hosts !!\n");
			printf("REDUZA O TAMANHO DA VM PARA %d\n",nndim);
			return;
		}
		kdif = 0;
		if(ncrit < nbase){
			printf("Voce tem menos casos do que o no. de hosts !!\n");
			kdif = NPROC - ncrit; // no. de casos artificiais para completar a carga
		}
		/*for(i=0; i<nrc; i++)
			jklasif[i] = klasif[i];*/

		/*for(i=kdif+ncrit; i<nrc; i++)
			jklasif[i] = -1;*/

		kount = 0;    // inicializa contador geral de casos
		icont0 = 0;   // inicializa contador de casos processados em "0"
		icontr = 0;   // inicializa contador de casos recebidos de outros hosts != "0"
/*---------------------- Manda os primeiros casos---------------------*/
		for(i=1; i<=nofinal; i++){
			//if(kount < ncrit){
				//kbusy[0] = jklasif[kount];
				kbusy[0] = klasif[kount];
				kbusy[1] = kount;
				kount++;
				icontr++;
				info = pvm_initsend(PvmDataDefault);
				info = pvm_pkint(kbusy,2,1);
				info = pvm_pkint(&ncrit,1,1);
				//info = pvm_pkint(klasif,ncrit,1);
				info = pvm_send(tids[i],msgtype2); // hostdest = i (i= 1,2,3,...,nofinal)
			//}
		}
		/*------No "0" pega 1 caso para analisar--------------*/
		if(kount < ncrit){
			casoAnalizado = klasif[kount];  // caso a ser analisado
			contingenciaAC(ce,cle,impr,casoAnalizado,kount);  // esta rotina faz chamadas da contingencia AC
			kount++;
			icont0++;
		}

		//while(casoAnalizado != -1){
		while(kount < ncrit + NPROC - 1  ){
			//----No "0" verifica o buffer---------
			knt = pvm_nrecv(-1,msgtype3);
			if(knt > 0){
				//printf("\n Verificando o Buffer");
				pvm_upkint(kbusy,2,1);
				ip = kbusy[0];    // tid do hostdest = ip
				nrcaso = kbusy[1];
				icontr++;
				//--Envia novo caso para "ip" analisar-------
				kbusy[0] = klasif[kount];
				kbusy[1] = kount;
				pvm_initsend( PvmDataDefault );
				pvm_pkint(kbusy, 2, 1);
				pvm_send(ip,msgtype4); // envia um caso para o escravo que terminou
				kount++;
			}
			else {
				if ( kount < ncrit){
		  //----- No "0" pega 1 caso para analisar--------
					casoAnalizado = klasif[kount];   // caso a ser analisado
					contingenciaAC(ce,cle,impr,casoAnalizado,kount);  // esta rotina faz chamadas da contingencia AC
					kount++;
					icont0++;
				}
			}
		}
	  /*--------- Fim do Processamento da Contingencia AC Lista Reduzida-
            ------------------------------------------------
	    Pega Resultados da Simulacao de Contingencias do
	    Processador "0"
	    -----------------------------------------------*/
		nilhasoma = nilha;
		ndivsoma = ndiv;
		nnconvsoma = nnconv;
		nyconvsoma = nyconv;
	/*------------------------------------------------
	 Recebe Resultados da Simulacao de Contingencias 
	 dos Demais Processadores
	 -----------------------------------------------
	// == Ilhamentos/Divergencias/Nao-Convergencias/Convergencias/Iteracoes: */
		printf("\n NO  | ILHAMENTO | DIVERG. | CONVERG.  | NAO-CONVERG. |\n");
		printf("\n %2d  | %5d     |%5d    |  %5d    |    %5d     |\n",myinst,nilha,ndiv,nyconv,nnconv);
	
		for(i=1; i<=nofinal; i++){
		//printf("recebendo os vetores");
		//info = -5;
		//while(info<=0)
		bufid = pvm_nrecv( tids[i], msgtype5 );
		while(bufid<=0)
			bufid = pvm_nrecv( tids[i], msgtype5 );

		info = pvm_upkint(&isend1[0],1,1 );
		info = pvm_upkint(&isend2[0],1,1 );
		info = pvm_upkint(&isend3[0],1,1 );
		info = pvm_upkint(&isend4[0],1,1 );

		nsize1[i] = isend1[0]+1;
		nsize2[i] = isend2[0]+1;
		nsize3[i] = isend3[0]+1;
		nsize4[i] = isend4[0]+1;
		}
	for(i=1; i<=nofinal; i++){
		nitens3 = nsize1[i];
		nitens4 = nsize2[i]; 
		nitens5 = nsize3[i];
		nitens6 = nsize4[i]; 
		info = pvm_recv( tids[i], msgtype6);
		info = pvm_upkint(isend1,nitens3,1);
		info = pvm_upkint(isend2,nitens4,1);
		info = pvm_upkint(isend3,nitens5,1);
		info = pvm_upkint(isend4,nitens6,1);
		info = pvm_upkint(isend5,nitens6,1);
		info = pvm_upkint(isend6,nitens6,1);
		info = pvm_upkint(isend7,nitens6,1);
		nilha = isend1[0];
		ndiv = isend2[0];
		nnconv = isend3[0];
		nyconv = isend4[0];
		//nyconv = isend5[0];
		//nyconv = isend6[0];
		// Recuperando kilha()
		if ( nilha != 0 ){
			for(j=0; j<nilha; j++)
				kilha[j+nilhasoma] = isend1[j+1];
			nilhasoma = nilhasoma + nilha;
		}
		// Recuperando kdiverge()
		if ( ndiv != 0 ){
			for(j=0; j<ndiv; j++)
				kdiverge[j+ndivsoma] = isend2[j+1];
			
			ndivsoma = ndivsoma + ndiv;
		}
		// Recuperando knotconv()
		if ( nnconv != 0 ){
			for(j=0; j<nnconv; j++)
				knotconv[j+nnconvsoma] = isend3[j+1];
			nnconvsoma = nnconvsoma + nnconv;
		}
		      // Recuperando kconv(), kit1() e kit2()
		if ( nyconv != 0 ){
			for(j=0; j<nyconv; j++){
				kconv[j+nyconvsoma] = isend4[j+1];
				kit1[j+nyconvsoma] = isend5[j+1];
				kit2[j+nyconvsoma] = isend6[j+1];
				kordem[j+nyconvsoma] = isend7[j+1];
			}
			nyconvsoma = nyconvsoma + nyconv;
		}
	}//fim for
	nilha = nilhasoma;
	ndiv = ndivsoma;
	nnconv = nnconvsoma;
	nyconv = nyconvsoma;
	} // if ( nbase )

	return;

}

void  MestreScreening(double ce[max],double cle[max],int impr){
	int ll,kk,k,l;
	/*---------------------------
	 Declaracao relativa ao PVM
	------------------------*/
	int info,bufid,aux;
	int icont0,icontr,nofinal;
	int nbase;
	int nitens1;
	int msgtype1;
	/* ---------------------------------------
	  Rotina MestreA(): Modo Paralelo Assincrono
	 ------------------------------------------
	 PVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVM
	--------------------------------------------
	--------------------------------
	Rotulos das mensagens [ types ]
	------------------------------*/
	msgtype1 = 21;
	nbase = NPROC;
	nofinal = NPROC-1;
	nitens1 = nrc + 1;
	/*---------------------------------
	Calculo do indice PI e ordenacao
	-------------------------------*/
	if( nbase == 1 ){
		/*--------------------------------------------
		Apenas o Processador "0" Executa [ serial ]
		------------------------------------------*/
		for(k=0; k<nrc; k++){
			//ioutlocal = ll; // caso a ser analisado
			scrdireto_reativo(ce,cle,impr,k);// esta rotina faz chamadas do metodo 1P-1Q
		}
	}
	else{
		/*-------------------------------------------
		Todos os Processadores Executam [ paralelo ]
		-------------------------------------------*/
		icont0 = 0;// inicializa contador de casos calculados pelo processador "0"
		for(k=myinst; k<nrc; k=k+nbase){
			//ioutlocal = kk;  // caso a ser analisado
			scrdireto_reativo(ce,cle,impr,k);// esta rotina faz chamadas do metodo 1P-1Q
			icont0++;
		}
		/*----------------------------------------------------------------
		1. Pega Resultados do  screening do Processador "0"
		2. Recebe Resultados dos Demais Processadores
		--------------------------------------------------------------*/
		for(k=1; k<=nofinal; k++){
			bufid = pvm_recv(tids[k],msgtype1);
			info = pvm_upkint(isend1,nitens1,1);
			info = pvm_upkint(&aux,1,1);
			info = pvm_upkdouble(psend,aux,1);
			icontr = aux-1;
			if ( icontr != 0 ){
				for(l=1; l<=icontr; l++){
					kk = isend1[l];  // no. do caso
					pia[kk] = psend[l];  // indice de severidade
				}
			}
		}
	}  // if ( nbase )
	
	reordena();// reordena a lista de indices de severidade
}
/*------------------ Fim da funcao MestreA-------------------*/

File: estagio/conting/PIPELINE/mestre.f
c23456781234567812345678123456781234567812345678123456781234567812345678
c   1       2       3       4       5       6       7       8       9
c ----------------------------------------------------------------------
c
c ------------------------------
c [ PROGRAMA PCX  == SERVIDOR ]
c ------------------------------

      subroutine MestreA(ce,impr,myinst,NPROC1,NPROC2,
     &NPROC3,NPROC,indice1,indice2,tids)
c
      include '/opt/pvm3/include/fpvm3.h'
c
      include 'globvar.inc'
      include 'varcont.inc'
      include 'varsel.inc'
      include 'tempos.inc'

c
c Cadeias de Zollenkopf
c
      dimension ce(max)
      integer*4 impr

c
c Declaracao Relativa ao PVM
c
      integer*4 myinst,info
      integer*4 NPROC,NPROC1,NPROC2,NPROC3
      integer*4 indice1(20),indice2(20),tids(20)


      dimension nsize2(63)                                  ! local
      dimension jklasif(nlmax)                              ! local


c ---------------------------------------------------
c Rotina MestreS(): Modo Paralelo Sincrono
c
c Rotina MestreA() e Pai(): Modo Paralelo Assincrono
c ---------------------------------------------------


c -----------------------------------------------
c  PVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVM
c -----------------------------------------------

c ------------------------------
c Rotulos das mensagens [types]
c ------------------------------
      nsg1 = 21
      nsg2 = 31
      nsg3 = 41
      nsg4 = 51
      nsg5 = 61
      nsg6 = 71



c ----------------------------------------------------------------------      
c                    1.] CALCULA DC INCREMENTAL 
c ----------------------------------------------------------------------


c ------------------------------------------------------
c          PROCESSAMENTO PARALELO ASSINCRONO
c ------------------------------------------------------
c ------------------------------------------------------
c O LOOP DC e Executado em Paralelo Assincrono: O  Pro-
c cessador Assim que Termina seu Caso Pega um Novo Caso
c para Executar 
c ------------------------------------------------------
c O Controle eh Feito Somente pelo Processo de  tids(1)   
c que Tambem Executa Calculos DC Enquanto Espera
c ------------------------------------------------------



c -------------------------------------
c Preparando a Carga dos Processadores
c -------------------------------------

      nbase = NPROC1

      if( nrc .lt. nbase ) then
          nndim = nrc
          write(*,*)' Toda a lista e menor do que o no. de hosts ! '
          write(*,*)' PCX) REDUZA O TAMANHO DA VM PARA ',nndim,' ! '
          return
      endif

      


      do i = 1, nrc
        jklasif(i) = i
      enddo
      jklasif(nrc+1) = 0  ! cartao flag



c ------------------------------------------------------------------------
c Situacao Em Que Se Dispoe de Um Unico Processador Para o DC Incremental
c ------------------------------------------------------------------------
      if( nbase .eq. 1 ) goto 950


      
       
c ------------------------------------------------------------------------------
c Situacao Em Que Se Dispoe de Dois ou Mais Processadores Para o DC Incremental
c ------------------------------------------------------------------------------

      kount = 0           ! inicializa o contador geral de casos
      icont0 = 0          ! inicializa o contador de casos processados em tids(1)
      icontr = 0          ! inicializa o contador de casos recebidos de outros hosts <> PCX Pai
 

c -------------------------
c Manda os primeiros casos 
c -------------------------

      do i = 1, NPROC1-1
        kp = i + 1       ! exclui o de tids(1)
        if ( kount .lt. nrc ) then
          kount = kount + 1
          kbusy(1) = jklasif(kount) 
          kbusy(2) = kount
          call PVMFINITSEND( PVMDEFAULT, info )
          call PVMFPACK( INTEGER4, kbusy, 2, 1, info )  
          call PVMFSEND( tids(kp), nsg2, info )   ! hostdest = i (i = 2, 3, ..., NPROC1)
        endif
      enddo
        
     
c ----------------------------------------------
c Processo de tids(1) pega 1 caso para analisar
c ----------------------------------------------

      if ( kount .lt. nrc ) then
        kount = kount + 1
        ioutlocal = jklasif(kount)                  ! caso a ser analisado

       	call contingenciaDC(ce,impr,myinst,NPROC1)  ! esta rotina faz chamadas do DC Incremental

        icont0 = icont0 + 1

c == Contabilidade:
         
        idp(kount) = tids(1)                        ! associa o caso analisado ao TID 
      endif

      

c ----------------------------------------
c Processo de tids(1) verifica o "buffer" 
c ----------------------------------------
      
    7 call PVMFNRECV( -1, nsg3, lnt )

      if ( lnt .gt. 0 ) then
        call PVMFUNPACK( INTEGER4, kbusy, 2, 1, info )

        ip = kbusy(1)                              ! tid do hostdest = ip
        nrcaso = kbusy(2)
        icontr = icontr + 1

c == Contabilidade:
         
        idp(nrcaso) = ip  ! associa o caso analisado ao TID 

c -----------------------------------
c Envia novo caso para "ip" analisar
c -----------------------------------

        kount = kount + 1              
        kbusy(1) = jklasif(kount) 
        kbusy(2) = kount

        call PVMFINITSEND( PVMDEFAULT, info )
        call PVMFPACK( INTEGER4, kbusy, 2, 1, info )  
        call PVMFSEND( ip, nsg4, info )    ! envia um caso para o escravo que terminou  

        if ( icont0 + icontr .eq. nrc ) goto 2000
        goto 7
      else

c ----------------------------------------------
c Processo de tids(1) pega 1 caso para analisar
c ----------------------------------------------

        if( kount .lt. nrc ) then
          kount = kount + 1
          ioutlocal = jklasif(kount)                 ! caso a ser analisado

       	  call contingenciaDC(ce,impr,myinst,NPROC1) ! esta rotina faz chamadas do DC Incremental

          icont0 = icont0 + 1

c == Contabilidade:
         
          idp(kount) = tids(1)                       ! associa o caso analisado ao TID 

          if ( icont0 + icontr .eq. nrc ) goto 2000
          goto 7
        endif
        goto 7
      endif                ! if ( lnt )

     
c -------------------------
c Termino do Processamento
c -------------------------  

 2000 continue             ! ioutlocal = 0 ==> sinal para parar




c -----------------------------------------------
c  PVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVM
c -----------------------------------------------

c ---------------------------------------------------------
c Pega Resultados do DC Incremental do Processo de tids(1)
c ---------------------------------------------------------
      nilhasoma = nilha

c ---------------------------------------------------------
c Recebe Resultados do DC Incremental dos Demais Processos
c ---------------------------------------------------------


c == Indices:

      do i = 1, NPROC1-1
        kp = i + 1                 ! exclui o de tids(1)
        call PVMFRECV( tids(kp), nsg1, bufid )
        call PVMFUNPACK( INTEGER4,isend1,nrc+1,1,info )
        call PVMFUNPACK( REAL8,psend,isend1(0),1,info )   
        
        iconti = isend1(0)
        if( iconti .ne. 0 ) then
           do j = 1, iconti
             kk = isend1(j)
             pia(kk) = psend(j)
           enddo
        endif
      enddo



      
c == Ilhamentos: 

ccc      write(*,*)' NO | ILHAMENTOS '
ccc      write(*,10)myinst,nilha
ccc   10 format(2x,i2,4x,i5,/)
     

      do i = 1, NPROC1-1
        kp = i + 1                       ! exclui o de tids(1)
        call PVMFRECV( tids(kp), nsg5, bufid )
        call PVMFUNPACK( INTEGER4, isend2(0),1,1,info )
        nsize2(kp) = isend2(0)+1
      enddo

      do i = 1, NPROC1-1
        kp = i + 1                       ! exclui o de tids(1)
        nitens4 = nsize2(kp) 

        call PVMFRECV( tids(kp), nsg6, bufid )
        call PVMFUNPACK( INTEGER4, isend2,nitens4,1,info )
        nilha = isend2(0)

       
c Recuperando kilha()
        if ( nilha .ne. 0 ) then
          do j = 1, nilha
            kilha(j+nilhasoma) = isend2(j)
          enddo
          nilhasoma = nilhasoma + nilha
        endif
      enddo

      nilha = nilhasoma



c ------------
c Da Um Salto
c ------------
      if( nbase .gt. 1 ) goto 1000
 
  


 
  950 continue
c ------------------------------------------------------------------------      
c                       1.] CALCULA DC INCREMENTAL 
c ------------------------------------------------------------------------
c ------------------------------------------------------------------------
c Situacao Em Que Se Dispoe de Um Unico Processador Para o DC Incremental
c ------------------------------------------------------------------------
      do i = 1, nrc

          ioutlocal = jklasif(i)

       	  call contingenciaDC(ce,impr,myinst,NPROC1) ! esta rotina faz chamadas do DC Incremental

c == Contabilidade:
         
        idp(i) = tids(1)                             ! associa o caso analisado ao TID

      enddo
c ------------------------------------------------------------------------



 1000 continue


      return
      end                       ! Fim da rotina MestreA()








c ----------------------------------------------------------------------
c
c ------------------------------
c [ PROGRAMA PCX  == SERVIDOR ]
c ------------------------------

      subroutine MestreS(ce,impr,myinst,NPROC1,NPROC2,
     &NPROC3,NPROC,indice1,indice2,tids)
c
      include '/opt/pvm3/include/fpvm3.h'
c
      include 'globvar.inc'
      include 'varcont.inc'
      include 'varsel.inc'
      include 'tempos.inc'

c
c Cadeias de Zollenkopf
c
      dimension ce(max)
      integer*4 impr

c
c Declaracao Relativa ao PVM
c
      integer*4 myinst,mytid,mynode,info
      integer*4 NPROC,NPROC1,NPROC2,NPROC3
      integer*4 indice1(20),indice2(20),tids(20)


      dimension nsize2(63)                                  ! local


c ---------------------------------------------------
c Rotina MestreS(): Modo Paralelo Sincrono
c
c Rotina MestreA() e Pai(): Modo Paralelo Assincrono
c ---------------------------------------------------


c -----------------------------------------------
c  PVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVM
c -----------------------------------------------

c ------------------------------
c Rotulos das mensagens [types]
c ------------------------------
      nsg1 = 21
      nsg2 = 31
      nsg3 = 41
      nsg4 = 51
      nsg5 = 61
      nsg6 = 71



c ----------------------------------------------------------------------      
c                    1.] CALCULA DC INCREMENTAL 
c ----------------------------------------------------------------------


c ------------------------------------------------
c        PROCESSAMENTO PARALELO SINCRONO
c ------------------------------------------------
c ------------------------------------------------
c O LOOP DC e Executado em Paralelo Sincrono: Os 
c Casos Sao Previamente Divididos Entre Hosts na
c Forma Balanceada em Numero de Casos
c ------------------------------------------------


c -----------------------------------
c Retorna o TID do processo corrente
c -----------------------------------

      call PVMFMYTID( mytid )

c -------------------------------------------------------------
c Define o "instance number" deste processo no contexto de PCX
c -------------------------------------------------------------

c -------------------------------------------------------------
c OBS: Isto eh feito aqui exclusivamente com o fim de permitir
c      o processamento paralelo SINCRONO (sem erros).
c -------------------------------------------------------------

      do i = 1, NPROC1
         if( mytid .eq. tids(i) ) mynode = i-1
      enddo

      nbase = NPROC1
c234567
      if( nbase .eq. 1 ) then

c ------------------------------------------------------------------------
c Situacao Em Que Se Dispoe de Um Unico Processador Para o DC Incremental
c ------------------------------------------------------------------------
      
c --------------------------------------------
c Apenas o Processador "0" Executa [ serial ]
c --------------------------------------------

	do ll = 1, nrc

          ioutlocal = ll                              ! caso a ser analisado

          call contingenciaDC(ce,impr,myinst,NPROC1)  ! esta rotina faz chamadas do DC Incremental

c == Contabilidade:
         
          idp(ioutlocal) = tids(1)                    ! associa o caso analisado ao TID

	enddo

      else          ! if ( nbase )      


c --------------------------------------------------------------------------
c Situacao Em Que Se Dispoe de Mais de Um Processador Para o DC Incremental
c --------------------------------------------------------------------------

       
c ---------------------------------------------
c Todos os Processadores Executam [ paralelo ]
c ---------------------------------------------

        icont0 = 0                                    ! inicializa contador de casos calculados pelo processador "0"

        do kk = mynode + 1, nrc, nbase

	  ioutlocal = kk                              ! caso a ser analisado

          call contingenciaDC(ce,impr,myinst,NPROC1)  ! esta rotina faz chamadas do DC Incremental

          icont0 = icont0 + 1

c == Contabilidade:
         
          idp(ioutlocal) = tids(1)                    ! associa o caso analisado ao TID
       
        enddo




c -----------------------------------------------
c  PVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVMPVM
c -----------------------------------------------

c ---------------------------------------------------------
c Pega Resultados do DC Incremental do Processo de tids(1) 
c ---------------------------------------------------------
      nilhasoma = nilha

c ---------------------------------------------------------
c Recebe Resultados do DC Incremental dos Demais Processos
c ---------------------------------------------------------


c == Indices:

      do i = 1, NPROC1-1
        kp = i + 1                 ! exclui o de tids(1)
        ktid = tids(kp)            ! TID do processo filho do qual se esta recebendo
        call PVMFRECV( tids(kp), nsg1, bufid )
        call PVMFUNPACK( INTEGER4,isend1,nrc+1,1,info )
        call PVMFUNPACK( REAL8,psend,isend1(0),1,info )   
        
        iconti = isend1(0)
        if( iconti .ne. 0 ) then
           do j = 1, iconti
             kk = isend1(j)
             pia(kk) = psend(j)

c == Contabilidade:
         
             idp(kk) = ktid       ! associa o caso analisado ao TID

           enddo
        endif
      enddo



      
c == Ilhamentos: 

ccc      write(*,*)' NO | ILHAMENTOS '
ccc      write(*,10)myinst,nilha
ccc   10 format(2x,i2,4x,i5,/)
     

      do i = 1, NPROC1-1
        kp = i + 1                       ! exclui o de tids(1)
        call PVMFRECV( tids(kp), nsg5, bufid )
        call PVMFUNPACK( INTEGER4, isend2(0),1,1,info )
        nsize2(kp) = isend2(0)+1
      enddo

      do i = 1, NPROC1-1
        kp = i + 1                       ! exclui o de tids(1)
        nitens4 = nsize2(kp) 

        call PVMFRECV( tids(kp), nsg6, bufid )
        call PVMFUNPACK( INTEGER4, isend2,nitens4,1,info )
        nilha = isend2(0)

       
c Recuperando kilha()
        if ( nilha .ne. 0 ) then
          do j = 1, nilha
            kilha(j+nilhasoma) = isend2(j)
          enddo
          nilhasoma = nilhasoma + nilha
        endif
      enddo

      nilha = nilhasoma


      endif          ! if ( nbase )      

      return
      end                       ! Fim da rotina MestreS()
